### PR TITLE
project: refactor storeapi

### PR DESCRIPTION
**Pull request by daniellim2000**
Refactor the 3 classes in `__init.py__` in the snapcraft.storeapi package into 3 different modules and  expose those modules in `__init__.py`

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

### CODE REVIEW

**Comment by daniellim2000** [comment]
Sorry I messed up the commits.

**Comment by daniellim2000** [comment]
I think because I accidentally made the previous commit in master it is now included here. I hope it does not pose an issue.

File: snapcraft/storeapi/__init__.py [hunk 1]
```diff
@@ -80,809 +53,15 @@ def _deserialize_macaroon(value):
...
-                content = requests.get(self.__status_details_url).json()
-            except (requests.ConnectionError, requests.HTTPError) as e:
-                if not connection_errors_allowed:
-                    yield e
-                content = {'processed': False, 'code': 'being_processed'}
-                connection_errors_allowed -= 1
-            yield content
+from .client import Client  # noqa
+from .store_client import (  # noqa
+    StoreClient,
```
Comment (sergiusens) [inline snapcraft/storeapi/__init__.py]: this should be the only thing in __init__, `StoreClient`

File: snapcraft/storeapi/__init__.py [hunk 2]
```diff
@@ -80,809 +53,15 @@ def _deserialize_macaroon(value):
...
-        while True:
-            try:
-                content = requests.get(self.__status_details_url).json()
-            except (requests.ConnectionError, requests.HTTPError) as e:
-                if not connection_errors_allowed:
-                    yield e
-                content = {'processed': False, 'code': 'being_processed'}
-                connection_errors_allowed -= 1
-            yield content
+from .client import Client  # noqa
```
Comment (sergiusens) [inline snapcraft/storeapi/__init__.py]: can you remove the .client here and make `Client` live in a module named `_client`?

File: snapcraft/storeapi/__init__.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/storeapi/__init__.py]: the module should also be named `._store_client`

File: snapcraft/storeapi/__init__.py [hunk 3]
```diff
@@ -80,809 +53,15 @@ def _deserialize_macaroon(value):
...
+from .store_client import (  # noqa
+    StoreClient,
+    SSOClient,
```
Comment (sergiusens) [inline snapcraft/storeapi/__init__.py]: can each of these live in there own module, also not import in `__init__`,
so
```
_sso_client -> SSOClient
_snap_index_client -> SnapIndexClient
```
and so on.
Also only import these where needed.

File: snapcraft/storeapi/__init__.py [hunk 4]
```diff
@@ -80,809 +53,15 @@ def _deserialize_macaroon(value):
...
+    SnapIndexClient,
+    UpDownClient,
+    SCAClient
+)
+from .status_tracker import (  # noqa
```
Comment (sergiusens) [inline snapcraft/storeapi/__init__.py]: can you make this private as well?

Comment (sergiusens) [reply snapcraft/storeapi/__init__.py]: and remove from here.

File: snapcraft/storeapi/__init__.py (hunk 1 above)
Comment (sergiusens) [reply snapcraft/storeapi/__init__.py]: It is also fine to import `errors` in this `__init__.py`

File: snapcraft/storeapi/store_client.py [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+from snapcraft.internal.indicators import download_requests_stream
+
+from .client import Client
+from .status_tracker import StatusTracker
+
+from . import logger, _macaroon_auth
+from . import _upload
+from . import constants
+from . import errors
+
```
Comment (sergiusens) [inline snapcraft/storeapi/store_client.py]: in here you would eventually do something like

```
from ._sso_client import SSOClient
```

and so on and so forth.

File: snapcraft/storeapi/__init__.py (hunk 1 above)
Comment (daniellim2000) [reply snapcraft/storeapi/__init__.py]: Do you mean `_store_client`?

Comment (daniellim2000) [reply snapcraft/storeapi/__init__.py]: If I don't import the other clients the tests won't pass

File: snapcraft/storeapi/__init__.py (hunk 3 above)
Comment (daniellim2000) [reply snapcraft/storeapi/__init__.py]: If I don't import the other clients the tests won't pass

File: snapcraft/storeapi/__init__.py (hunk 4 above)
Comment (daniellim2000) [reply snapcraft/storeapi/__init__.py]: The tests also need progressbar to pass

**Comment by sergiusens** [comment]
Is it possible to adapt the tests for this to work?
Cheers

**Comment by daniellim2000** [comment]
Yea no problem I'm working on it, but busy these few days so may take some time.

**Comment by daniellim2000** [comment]
Sorry I realized there are still some problems with the push.

**Comment by daniellim2000** [comment]
I kept getting
`ValueError: Timeout value connect was Timeout(connect=None, read=None, total=None), but it must be an int or float.`

Is it due to my network connection?

**Comment by daniellim2000** [comment]
Regarding the error above, it turns out that there suddenly was a problem with my requests module, for whatever reason. Reinstalling it fixed the problem for me.

File: snapcraft/storeapi/_store_client.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    def __init__(self):
+        super().__init__()
+        self.conf = config.Config()
+        self.sso = SSOClient(self.conf)
+        self.cpi = SnapIndexClient(self.conf)
+        self.updown = UpDownClient(self.conf)
+        self.sca = SCAClient(self.conf)
+
+    def login(self, email, password, one_time_password=None, acls=None,
+              packages=None, channels=None, save=True):
```
Comment (kalikiana) [inline snapcraft/storeapi/_store_client.py]: It seems the mypy annotions have gotten lost in the rewrite? ie. on master the function has this signature:

        def login(self, email: str, password: str, one_time_password: str = None,
              acls: Iterable[str] = None, channels: Iterable[str] = None,
              packages: Iterable[Dict[str, str]] = None,
              config_fd: TextIO = None, save: bool = True) -> None:

Can you please add these again? They help spot issues with different types, shadowed variables and such early as part of `./runtests.sh static`.

File: snapcraft/storeapi/_sca_client.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+            })
+        if channels is not None:
+            data.update({
+                'channels': channels,
+            })
+        headers = {
+            'Accept': 'application/json',
+        }
+        response = self.post(
+            'acl/', json=data, headers=headers)
```
Comment (kalikiana) [inline snapcraft/storeapi/_sca_client.py]: Compared to the old code, it seems the `acl/verify` endpoint of the API isn't used anymore? Is this done differently now or was it redundant? I'm not that familiar with this code, so I might be missing something here. Can you please clarify?

File: snapcraft/storeapi/_store_client.py (hunk 6 above)
Comment (daniellim2000) [reply snapcraft/storeapi/_store_client.py]: Sorry I'm not sure what happened there as well. Changing it now.

File: snapcraft/storeapi/_sca_client.py (hunk 7 above)
Comment (daniellim2000) [reply snapcraft/storeapi/_sca_client.py]: I think perhaps that was changed some time ago?

File: snapcraft/storeapi/_store_client.py (hunk 6 above)
Comment (kalikiana) [reply snapcraft/storeapi/_store_client.py]: Would you mind also adding ` -> None` to __init__? I'm afraid I only spotted this now.

**Comment by daniellim2000** [comment]
Alright I think after creating this branch there were some updates made to master that I did not sync, so some of the stuff are outdated.

I've re-copied everything from master.

**Comment by daniellim2000** [comment]
Thanks!

**Pull request merged**